Orm/Xtensive.Orm.PostgreSql/Sql.Drivers.PostgreSql/v8_0/Translator.cs

```diff
@@ -67,86 +67,115 @@ protected override void TranslateChar(IOutput output, char ch)
+    {
+      switch(type) {
+        case SqlFunctionType.SystemUser: return;
+        case SqlFunctionType.User:
+        case SqlFunctionType.CurrentUser:
+          _ = output.Append("current_user");
+          break;
+        case SqlFunctionType.SessionUser: _ = output.Append("session_user"); break;
+        case SqlFunctionType.NullIf: _ = output.Append("nullif"); break;
+        case SqlFunctionType.Coalesce: _ = output.Append("coalesce"); break;
+        case SqlFunctionType.BinaryLength: _ = output.Append("length"); break;
+        //datetime/timespan
+        case SqlFunctionType.CurrentDate: _ = output.Append("date_trunc('day', current_timestamp)"); break;
+        case SqlFunctionType.CurrentTimeStamp: _ = output.Append("current_timestamp"); break;
+        case SqlFunctionType.IntervalNegate: _ = output.Append("-"); break;
+        //string
+        case SqlFunctionType.CharLength: _ = output.Append("char_length"); break;
+        case SqlFunctionType.Lower: _ = output.Append("lower"); break;
+        case SqlFunctionType.Position: _ = output.Append("position"); break;
+        case SqlFunctionType.Substring: _ = output.Append("substring"); break;
+        case SqlFunctionType.Upper: _ = output.Append("upper"); break;
+        case SqlFunctionType.Concat: _ = output.Append("textcat"); break;
+        //math
+        case SqlFunctionType.Abs: _ = output.Append("abs"); break;
+        case SqlFunctionType.Acos: _ = output.Append("acos"); break;
+        case SqlFunctionType.Asin: _ = output.Append("asin"); break;
+        case SqlFunctionType.Atan: _ = output.Append("atan"); break;
+        case SqlFunctionType.Atan2: _ = output.Append("atan2"); break;
+        case SqlFunctionType.Ceiling: _ = output.Append("ceil"); break;
+        case SqlFunctionType.Cos: _ = output.Append("cos"); break;
+        case SqlFunctionType.Cot: _ = output.Append("cot"); break;
+        case SqlFunctionType.Degrees: _ = output.Append("degrees"); break;
+        case SqlFunctionType.Exp: _ = output.Append("exp"); break;
+        case SqlFunctionType.Floor: _ = output.Append("floor"); break;
+        case SqlFunctionType.Log: _ = output.Append("ln"); break;
+        case SqlFunctionType.Log10: _ = output.Append("log"); break;
+        case SqlFunctionType.Pi: _ = output.Append("pi"); break;
+        case SqlFunctionType.Power: _ = output.Append("power"); break;
+        case SqlFunctionType.Radians: _ = output.Append("radians"); break;
+        case SqlFunctionType.Rand: _ = output.Append("random"); break;
+        case SqlFunctionType.Round: _ = output.Append("round"); break;
+        case SqlFunctionType.Truncate: _ = output.Append("trunc"); break;
+        case SqlFunctionType.Sign: _ = output.Append("sign"); break;
+        case SqlFunctionType.Sqrt: _ = output.Append("sqrt"); break;
+        case SqlFunctionType.Tan: _ = output.Append("tan"); break;
+        default: base.Translate(output, type); break;
```

Comment:
It would be better to make this translation vis `Dictionary<SqlFunctionType, string>`
In the form of huge switch there is much of code, which pollutes CPU cache

Reply:
Dictionary is slow comparing with pure switch. If we spend some memory for the sake of better CPU cache utilization I did it by using an lightweight type based on array. As shown below, array of 66 items spend less time on population and less memory that dictionary of 37 items (which we need here). Yes, we waste a little memory for unused indexes but as it's shown not so much
```
|             Method |      Mean |    Error |   StdDev | Rank |  Gen 0 |  Gen 1 | Allocated |
|------------------- |----------:|---------:|---------:|-----:|-------:|-------:|----------:|
|        ArrayFillIn |  78.04 ns | 0.667 ns | 0.623 ns |    1 | 0.1173 |      - |     552 B |
|         DictFillIn | 569.36 ns | 3.521 ns | 3.293 ns |    3 | 0.4416 | 0.0019 |   2,080 B |
| TranslationsFillIn |  84.72 ns | 1.346 ns | 1.124 ns |    2 | 0.1223 | 0.0001 |     576 B |
```
Here's some benchmark results of translation for enum values from different parts of original switch:
```
|           Method |        Type |     Mean |     Error |    StdDev | Rank | Allocated |
|----------------- |------------ |---------:|----------:|----------:|-----:|----------:|
|       Dictionary | CurrentDate | 9.157 ns | 0.0106 ns | 0.0099 ns |    9 |         - |
|           Switch | CurrentDate | 1.867 ns | 0.0038 ns | 0.0035 ns |    4 |         - |
| SwitchExpression | CurrentDate | 1.864 ns | 0.0027 ns | 0.0021 ns |    4 |         - |
|            Array | CurrentDate | 1.548 ns | 0.0026 ns | 0.0024 ns |    3 |         - |
|     Translations | CurrentDate | 1.490 ns | 0.0026 ns | 0.0022 ns |    2 |         - |
|       Dictionary |  CharLength | 9.230 ns | 0.0151 ns | 0.0134 ns |    9 |         - |
|           Switch |  CharLength | 1.865 ns | 0.0017 ns | 0.0014 ns |    4 |         - |
| SwitchExpression |  CharLength | 1.869 ns | 0.0029 ns | 0.0026 ns |    4 |         - |
|            Array |  CharLength | 1.294 ns | 0.0024 ns | 0.0022 ns |    1 |         - |
|     Translations |  CharLength | 1.493 ns | 0.0031 ns | 0.0029 ns |    2 |         - |
|       Dictionary | CurrentUser | 9.045 ns | 0.0139 ns | 0.0130 ns |    8 |         - |
|           Switch | CurrentUser | 2.150 ns | 0.0026 ns | 0.0024 ns |    6 |         - |
| SwitchExpression | CurrentUser | 2.148 ns | 0.0018 ns | 0.0016 ns |    6 |         - |
|            Array | CurrentUser | 1.291 ns | 0.0022 ns | 0.0019 ns |    1 |         - |
|     Translations | CurrentUser | 1.490 ns | 0.0017 ns | 0.0015 ns |    2 |         - |
|       Dictionary |        User | 8.294 ns | 0.0117 ns | 0.0109 ns |    7 |         - |
|           Switch |        User | 2.149 ns | 0.0043 ns | 0.0038 ns |    6 |         - |
| SwitchExpression |        User | 2.151 ns | 0.0047 ns | 0.0044 ns |    6 |         - |
|            Array |        User | 1.293 ns | 0.0026 ns | 0.0025 ns |    1 |         - |
|     Translations |        User | 1.495 ns | 0.0055 ns | 0.0052 ns |    2 |         - |
|       Dictionary |         Cot | 8.235 ns | 0.0111 ns | 0.0099 ns |    7 |         - |
|           Switch |         Cot | 2.119 ns | 0.0023 ns | 0.0021 ns |    5 |         - |
| SwitchExpression |         Cot | 2.117 ns | 0.0016 ns | 0.0013 ns |    5 |         - |
|            Array |         Cot | 1.293 ns | 0.0024 ns | 0.0021 ns |    1 |         - |
|     Translations |         Cot | 1.493 ns | 0.0037 ns | 0.0033 ns |    2 |         - |
```

The type performance is really close to array and faster than rest of variants

Reply:
I totally agree with this optimization.
It may be even more fast it you remove one level of inderection, converting `SqlFunctionTypeTranslations` into struct